std/machines/hash/poseidon2_bb.asm

```diff
@@ -0,0 +1,288 @@
+use std::array;
+use std::check::assert;
+use std::utils;
+use std::utils::unchanged_until;
+use std::utils::force_bool;
+use std::utils::sum;
+use std::convert::expr;
+use std::machines::small_field::memory::Memory;
+use std::machines::small_field::pointer_arith::increment_ptr;
+use std::machines::split::split_bb::SplitBB;
+
+// Implements the Poseidon2 permutation for BabyBear field.
+//
+// Apparently it can be used to hash arbitrary sized data by using the
+// Merkle–Damgård construction, or it can be used as a compression function
+// for building a Merkle tree.
```

Comment:
> Hmm, so the function permutes a 512-Bit input and returns the first 256 Bit, right?

Right. This is the version used by p3 backend, so it should be useful for recursion.

> This is different from the GL Poseidon, which permutes a 768-Bit input (12 64-Bit field elements) and returns the first 256 Bit. There, you'd input your 512-Bit input (e.g. the two child hashes in a Merkle tree) plus a 256-Bit capacity, which would be a constant (where different constants give you different hash functions).

Yes.

> Are you sure we're not supposed to put in a 256-Bit constant here as well? How would we get different hash functions (e.g. one for the Merkle leaves and one for the internal nodes)?

For internal nodes a full hash function is not necessary, just a compression is sufficient. Which this is.

For the leaves, I don't know for sure. How do you do with GL?

I can see 2 options. Alternative a) is Merkle-tree all the way down to 512 bits of data (it works because the data is a fixed power of 2 size). I guess this is safe because this is an intended usage in Poseidon2 paper.

Alternative b) is to use a construction that can turn a compression function into a hash function, like Merkle–Damgård (or its updated version, HAIFA construct). This usage is just implied in passing in Poseidon2 paper. If you need 2^256 different hash functions, set the IV to something other than zero in the first compression step, but I don't see why it would be needed in the bootloader.

Alternative a) seems simpler to me.

Reply:
> For internal nodes a full hash function is not necessary, just a compression is sufficient. Which this is.

What do you mean by "compression is sufficient"? I think we need collision resistance, right? I would *guess* that this construction would be collision resistant, I'm just wondering if that's a standard implementation or whether we do something unusual here.

And yes, for the bootloader the Merkle tree in the bootloader has a pre-defined depth, so it is fine to use the same hash function!

Reply:
This is a quote from Poseidon2 paper that I interpret as meaning it is safe to use a single Poseidon2 permutation to level-up a Merkle-tree:

> While hash functions are perfectly usable and allow to make strong security
arguments, they are often more generic. For many classical applications (e.g.,
ordinary hashing of arbitrary-length inputs) this is not a major issue. However,
when building a Merkle tree with small fixed-size input lengths, we often only
need a single permutation call to process the entire input and compute the
desired output. In this case, the construction is not used as a hash function in a
traditional sense, but rather as a compression function.